services/events/ingestion.go

```diff
@@ -0,0 +1,192 @@
+package events
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-evm-gateway/models"
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-go-sdk"
+	"github.com/rs/zerolog"
+)
+
+var ErrDisconnected = errors.New("disconnected")
+
+var _ models.Engine = &EventIngestionEngine{}
+
+type EventIngestionEngine struct {
+	subscriber   Subscriber
+	blocks       storage.BlockIndexer
+	receipts     storage.ReceiptIndexer
+	transactions storage.TransactionIndexer
+	logs         zerolog.Logger
+	lastHeight   *models.SequentialHeight
+	status       *models.EngineStatus
+}
+
+func NewEventIngestionEngine(
+	subscriber Subscriber,
+	blocks storage.BlockIndexer,
+	receipts storage.ReceiptIndexer,
+	transactions storage.TransactionIndexer,
+	logs zerolog.Logger,
+) *EventIngestionEngine {
+	return &EventIngestionEngine{
+		subscriber:   subscriber,
+		blocks:       blocks,
+		receipts:     receipts,
+		transactions: transactions,
+		logs:         logs,
+		status:       models.NewEngineStatus(),
+	}
+}
+
+// Ready signals when the engine has started.
+func (e *EventIngestionEngine) Ready() <-chan struct{} {
+	return e.status.IsReady()
+}
+
+// Done signals when the engine has stopped.
+func (e *EventIngestionEngine) Done() <-chan struct{} {
+	// return e.status.IsDone()
+	return nil
+}
+
+// Stop the engine.
+func (e *EventIngestionEngine) Stop() {
+	// todo
+}
+
+// Start the event ingestion engine. Load the latest height that was stored and provide it
+// to the event subscribers as a starting point.
+// Consume the events provided by the event subscriber.
+// Each event is then processed by the event processing methods.
+func (e *EventIngestionEngine) Start(ctx context.Context) error {
+	latest, err := e.blocks.LatestHeight()
```

Comment:
How would that work for "cold" starts? For example, a developer running this locally, will probably connect to an Emulator, where it should start from block height 0. Or connect to testnet/mainnet, where it should start from the block height where the `EVM` contract was first deployed.

Reply:
This startup and configuration will be part of bootstrap not part of the engine, because the database will have to be initialized, there it will be checked whether the db is initialized and whether the operator provided start height etc.

Reply:
Thanks for the explanation, that makes total sense :+1: